### PR TITLE
Add an example for setting and reading error details

### PR DESCRIPTION
**Pull request by chikei**
Mostly copied from [ErrorHandlingClient](https://github.com/grpc/grpc-java/blob/master/examples/src/main/java/io/grpc/examples/errorhandling/ErrorHandlingClient.java), or should I just modify `ErrorHandlingClient` ?
for #1295

### CODE REVIEW

File: examples/src/main/proto/debug_info.proto [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "io.grpc.examples.errorhandling";
+option java_outer_classname = "DebugInfoProto";
+option objc_class_prefix = "DEI";
+
+package debuginfo;
+
+message DebugInfo {
```
Comment (ejona86) [inline examples/src/main/proto/debug_info.proto]: Might be good to link to https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto, or copy the file. I think eventually they may be a library we can depend on.

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+/**
+ * Shows how to setting and reading RPC error details
+ */
+public class DetailErrorSample {
+  static final Metadata.Key<byte[]>
+    DEBUG_INFO_TRAILER_KEY = Metadata.Key.of("debug-info-bin", Metadata.BINARY_BYTE_MARSHALLER);
```
Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: This should probably be `ProtoUtils.keyForProto(DebugInfo.getDefaultInstance())`.

**Comment by ejona86** [comment]
It's probably not too bad to modify the current ErrorHandlingClient, as long as we have comments to make it obvious this is for additional details.

**Comment by ejona86** [comment]
@carl-mastrangelo, do you have an opinion?

**Comment by carl-mastrangelo** [comment]
Not really.  I misinterpreted the original bug for it to be that it was not clear how to do error handling, and didn't see that it was actually to implement the proto in the googleapis repo.  

Copying the code seems fine to me, but I would like to see the whole proto brought in, not just part.  It isn't obvious from the proto in this PR that it is the full-fledged, endorsed-by-google one.

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+/**
+ * Shows how to setting and reading RPC error details
+ */
+public class DetailErrorSample {
+  static final Metadata.Key<DebugInfo>
```
Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: You'll need to import DebugInfo now.

Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: Include a comment that this is just an example proto that adds additional information, and that it is also common to be an application-specific proto.

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  private Server server;
+  private ManagedChannel channel;
+
+  void run() throws Exception {
+    server = ServerBuilder.forPort(0).addService(new GreeterGrpc.GreeterImplBase() {
+      @Override
+      public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
+        Metadata trailers = new Metadata();
+        trailers.put(DEBUG_INFO_TRAILER_KEY, DEBUG_INFO);
+        responseObserver.onError(Status.INTERNAL.asRuntimeException(trailers));
```
Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: Still include a withDescription.

**Comment by ejona86** [comment]
@chikei LGTM. @carl-mastrangelo, could you review as well?

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * Shows how to setting and reading RPC error details.
+ * Proto used here is just an example proto, but the pattern sending
+ * application error information as an application-specific binary protos
+ * in the response trailers is the recommended way to return application
+ * level error.
+ */
+public class DetailErrorSample {
```
Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: Sorry for the flurry of nits.   Most of these happen automatically using Intellij or Eclipse.

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  static final Metadata.Key<DebugInfo>
```
Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: could you break the line at the equals?  We try and break lines at the highest semantic level to keep related tokens together.

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    DEBUG_INFO_TRAILER_KEY = ProtoUtils.keyForProto(DebugInfo.getDefaultInstance());
+
+  static final DebugInfo DEBUG_INFO =
```
Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: private

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    DebugInfo.newBuilder()
```
Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: We use +4 indent for line wrap, and +2 indent for new block.  Could you update the code in this PR to do this?

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+  static void verifyErrorReply(Throwable t) {
+    Status status = Status.fromThrowable(t);
+    Metadata trailers = Status.trailersFromThrowable(t);
+    Verify.verify(status.getCode() == Status.Code.INTERNAL);
+    Verify.verify(trailers.containsKey(DEBUG_INFO_TRAILER_KEY));
+    Verify.verify(status.getDescription().equals(DEBUG_DESC));
+    try {
+      Verify.verify(trailers.get(DEBUG_INFO_TRAILER_KEY).equals(DEBUG_INFO));
+    } catch (IllegalArgumentException i) {
```
Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: s/i/e/

Comment (carl-mastrangelo) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: Where does the IAE come from?

File: examples/src/main/proto/error_details.proto [hunk 10]
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2015, Google Inc.
```
Comment (carl-mastrangelo) [inline examples/src/main/proto/error_details.proto]: 2016 ?

File: examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java (hunk 9 above)
Comment (chikei) [inline examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java]: From the `parseBytes` of `ProtoLiteUtils.metadataMarshaller` inside the `ProtoUtils.keyForProto`, maybe I should just drop it ?

File: examples/src/main/proto/error_details.proto (hunk 10 above)
Comment (chikei) [inline examples/src/main/proto/error_details.proto]: Copied from the original file, not sure about your copyright section policy, I'll update it.

**Comment by carl-mastrangelo** [comment]
@chikei fyi we typically don't look at a review again until you ping the PR, typically with a "PTAL" comment.  Let us know when this is ready again to review.

**Comment by chikei** [comment]
@carl-mastrangelo ah, didn't know that, PTAL.

**Comment by carl-mastrangelo** [comment]
@chikei LGTM

**Pull request merged**